src/app/clusters/general-diagnostics-server/general-diagnostics-server.cpp

```diff
@@ -205,113 +205,118 @@ class GeneralDiagnosticsDelegate : public DeviceLayer::ConnectivityManagerDelega
+                DataModel::List<const RadioFaultType> previousList = DataModel::List<const RadioFaultType>(
+                    reinterpret_cast<const RadioFaultType *>(previous.data()), previous.size());
+                Events::RadioFaultChange::Type event{ currentList, previousList };
+
+                if (CHIP_NO_ERROR != LogEvent(event, endpointId, eventNumber, EventOptions::Type::kUrgent))
                 {
-                    // If General Diagnostics cluster is implemented on this endpoint
-                    MatterReportingAttributeChangeCallback(endpointId, GeneralDiagnostics::Id,
-                                                           GeneralDiagnostics::Attributes::ActiveRadioFaults::Id);
-
-                    // Record RadioFault event
-                    EventNumber eventNumber;
-                    DataModel::List<const RadioFaultType> currentList = DataModel::List<const RadioFaultType>(
-                        reinterpret_cast<const RadioFaultType *>(current.data()), current.size());
-                    DataModel::List<const RadioFaultType> previousList = DataModel::List<const RadioFaultType>(
-                        reinterpret_cast<const RadioFaultType *>(previous.data()), previous.size());
-                    Events::RadioFaultChange::Type event{ currentList, previousList };
-
-                    if (CHIP_NO_ERROR != LogEvent(event, endpointId, eventNumber, EventOptions::Type::kUrgent))
-                    {
-                        ChipLogError(Zcl, "GeneralDiagnosticsDelegate: Failed to record RadioFault event");
-                    }
+                    ChipLogError(Zcl, "GeneralDiagnosticsDelegate: Failed to record RadioFault event");
                 }
-            }
-        }
+
+                return Loop::Continue;
+            },
+            reinterpret_cast<intptr_t>(&faults));
     }
 
     // Get called when the Node detects a network fault has been raised.
-    void OnNetworkFaultsDetected(GeneralFaults<kMaxNetworkFaults> & previous, GeneralFaults<kMaxNetworkFaults> & current) override
+    void OnNetworkFaultsDetected(GeneralFaults<kMaxNetworkFaults> & previousArg,
+                                 GeneralFaults<kMaxNetworkFaults> & currentArg) override
     {
         ChipLogProgress(Zcl, "GeneralDiagnosticsDelegate: OnHardwareFaultsDetected");
 
-        ReportAttributeOnAllEndpoints(GeneralDiagnostics::Attributes::ActiveNetworkFaults::Id);
-
-        for (uint16_t index = 0; index < emberAfEndpointCount(); index++)
-        {
-            if (emberAfEndpointIndexIsEnabled(index))
-            {
-                EndpointId endpointId = emberAfEndpointFromIndex(index);
-
-                if (emberAfContainsServer(endpointId, GeneralDiagnostics::Id))
+        using Faults  = TFaults<kMaxNetworkFaults>;
+        Faults faults = { previousArg, currentArg };
+        ForAllEndpointsWithServerCluster(
```

Comment:
idea for "the future": Perhaps an alternative approach using endpoint iterators would help avoid reinterpret_casts. Something like:
```
class EndpointRange
{
public:
  EndpointRange(ClusterId clusterId) : mBegin(clusterId, 0), mEnd(clusterId, emberAfEndpointCount()) {}
  auto begin() const { return mBegin; }
  auto end() const { return mEnd; }

private:
  EndpointIterator mBegin;
  EndpointIterator mEnd;
};

for (auto endpointId : EndpointRange(GeneralDiagnostics::Id)) ...
```

Reply:
I'll take a look and see how the codesize looks.  The code for `ForAllEndpointsWithServerCluster` ends up being smaller than the manual loops....

Reply:
Yes, that's a good point.